archive/imo/imo2013_q5.lean

```diff
@@ -0,0 +1,413 @@
+   {
+     intros n hn,
+     have hterm : (∀i:ℕ, i ∈ finset.range n → 1 ≤ x^i * y^(n - 1 - i)),
+     {
+       intros i hi,
+       have hx' : 1 ≤ x ^ i := one_le_pow_of_one_le (le_of_lt hx) i,
+       have hy' : 1 ≤ y ^ (n - 1 - i) := one_le_pow_of_one_le (le_of_lt hy) (n - 1 - i),
+       nlinarith,
+     },
+     have : ((n:ℝ) = (∑ (i : ℕ) in finset.range n, (1:ℝ))),
+     {
+       simp only [mul_one, finset.sum_const, nsmul_eq_mul, finset.card_range]
+     },
+
+     calc (x - y) * (n:ℝ)
+             = (x - y) *  (∑ (i : ℕ) in finset.range n, (1:ℝ))
+               : by simp only [mul_one, finset.sum_const, nsmul_eq_mul, finset.card_range]
+         ... ≤ (x-y) * (∑ (i : ℕ) in finset.range n, x ^ i * y ^ (n - 1 - i))
+               : (mul_le_mul_left hxmy).mpr (finset.sum_le_sum hterm)
+         ... = x^n - y^n : (factor_xn_minus_yn x y n).symm,
+   },
+
+   -- Choose n larger than 1 / (x - y).
+   obtain ⟨N, hN⟩ := exists_nat_gt (1 / (x - y)),
+   have hNpr : (0:ℝ) < N := lt_trans (one_div_pos.mpr hxmy) hN,
+   have hNp : 0 < N,
+   {
+      cases N,
+      { exfalso, exact lt_asymm hNpr hNpr },
+      exact nat.succ_pos N
+   },
+
+   have := calc 1 = (x - y) * (1 / (x - y)) : by field_simp [ne_of_gt hxmy]
+              ... < (x - y) * N             : (mul_lt_mul_left hxmy).mpr hN
+              ... ≤ x^N - y^N               : hn N hNp,
+   linarith [h N hNp],
+end
+
+/--
+ Like le_of_all_pow_lt_succ, but with a weaker assumption for y.
+-/
+lemma le_of_all_pow_lt_succ' (x y: ℝ) (hx: 1 < x) (hy: 0 < y)
+      (h: ∀n:ℕ, 0 < n → x^n - 1 < y^n)
+      : x ≤ y :=
+begin
+  have hy': 1 < y,
+  {
+    by_contra hy'',
+    push_neg at hy'', -- hy'' :  y ≤ 1.
+
+    -- Then there exists y' such that 0 < y ≤ 1 < y' < x.
+    let y' := (x + 1) / 2,
+    have h_y'_lt_x: y' < x,
+    {
+      have hh: (x + 1)/2 < (x * 2) / 2 := by linarith,
+      calc y' < (x * 2) / 2 : hh
+          ... = x : by field_simp
+    },
+    have h1_lt_y' : 1 < y',
+    {
+      have hh': 1 * 2 / 2 < (x + 1) / 2 := by linarith,
+      calc 1 = 1 * 2 / 2 : by field_simp
+         ... < y' : hh'
+    },
+    have h_y_lt_y' := calc y ≤ 1 : hy''
+                         ... < y': h1_lt_y',
+    have hh: (∀ n, 0 < n → x^n - 1 < y'^n),
+    {
+      intros n hn,
+      calc x^n - 1 < y^n : h n hn
+              ...  ≤ y'^n : pow_le_pow_of_le_left (le_of_lt hy) (le_of_lt h_y_lt_y') n
+    },
+    have : x ≤ y' := le_of_all_pow_lt_succ x y' hx h1_lt_y' hh,
+    linarith, -- contradiction
+  },
+  exact le_of_all_pow_lt_succ x y hx hy' h
+end
+
+lemma twice_pos_int_gt_one (z: ℤ) (hpos: 0 < z) : (1:ℚ) < (((2 * z):ℤ):ℚ) :=
+by { norm_cast, linarith }
+
+theorem imo2013_q5
+  (f: ℚ → ℝ)
+  (H1:  ∀ x y, 0 < x → 0 < y → f (x * y) ≤ f x * f y)
+  (H2: ∀ x y, 0 < x → 0 < y → f x + f y ≤ f (x + y))
+  (H_fixed_point: ∃ a, 1 < a ∧ f a = a)
+  : ∀ x, 0 < x → f x = x :=
+begin
+  obtain ⟨a, ha1, hae⟩ := H_fixed_point,
+  have H3: ∀x: ℚ, (0 < x → ∀ n: ℕ, 0 < n → ↑n * f x ≤ f (n * x)),
```

Comment:
I think some of these `have`s could be extracted to separate lemmas. To avoid restating the same hypotheses over and over, you can use `variables` and `section`.